### PR TITLE
Committee contract unit test cases

### LINKED ISSUE #33: Unit test cases for committee contract methods

### PR DESCRIPTION
**Pull request by gulshanvasnani**
Do not merge before https://github.com/mosaicdao/mosaic-1/pull/25
PR contains changes to add negative unit test cases for committee status.

fixes #33

### CODE REVIEW

File: test/committee/utils.js [hunk 1]
```diff
@@ -19,6 +19,28 @@ const web3 = require('../test_lib/web3.js');
 
 const Committee = artifacts.require('Committee');
 
+function getMemberDistance(accountProvider, dislocation, proposal, committeeSize, compareFn) {
+  const dist = [];
+  for (let i = 0; i < committeeSize + 2; i += 1) {
```
Comment (pgev) [inline test/committee/utils.js:24]: What is "+2" here for?

File: test/committee/utils.js [hunk 2]
```diff
@@ -19,6 +19,28 @@ const web3 = require('../test_lib/web3.js');
...
+    const account = accountProvider.get();
+    dist.push({
+      address: account,
+      distance: this.distanceToProposal(
```
Comment (pgev) [inline test/committee/utils.js]: What is `this` in this case?

File: test/committee/utils.js [hunk 3]
```diff
@@ -19,6 +19,28 @@ const web3 = require('../test_lib/web3.js');
 
 const Committee = artifacts.require('Committee');
 
+function getMemberDistance(accountProvider, dislocation, proposal, committeeSize, compareFn) {
```
Comment (pgev) [inline test/committee/utils.js]: Proposal:
  * getMemberDistance -> getCommitteeMembers
  * compareFn -> sortPredicate

File: test/committee/utils.js [hunk 4]
```diff
@@ -19,6 +19,28 @@ const web3 = require('../test_lib/web3.js');
...
+        proposal,
+      ),
+    });
+  }
+
+  dist.sort(compareFn);
+  return dist;
+}
+
+function compare(a, b) {
```
Comment (pgev) [inline test/committee/utils.js]: Proposal:
  * compareMemberDistance(firstMember, secondMember)

File: test/committee/activate_committee.js [hunk 5]
```diff
@@ -85,7 +81,7 @@ contract('Committee::activateCommittee', async (accounts) => {
       );
     });
 
-    it('should fail if committee is not in cooling down mode', async () => {
+    it('should fail if committee is in open phase status', async () => {
```
Comment (pgev) [inline test/committee/activate_committee.js:84]: As it has been decided to assert the condition in all other states, we are missing "CommitteeStatus::Closed" state here.

File: test/committee/challenge_committee.js [hunk 6]
```diff
@@ -79,15 +69,15 @@ contract('Committee:challengeCommittee', async (accounts) => {
...
+        from: config.committee.member,
       },
     );
 
     Object.freeze(config);
   });
 
   contract('Negative Tests', async () => {
-    it('should fail if committee is not in cooling down mode', async () => {
+    it('should fail if committee is in open phase state ', async () => {
```
Comment (pgev) [inline test/committee/challenge_committee.js:80]: As it has been decided to assert the condition in all other states, we are missing `CommitteeStatus::Closed` and `CommitteeStatus::Invalid` states here.

File: test/committee/close_commit_phase.js [hunk 7]
```diff
@@ -49,50 +49,124 @@ contract('Committee::closeCommitPhase', async (accounts) => {
...
   });
 
   contract('Negative Tests', async () => {
-    it('should fail if committee is not in commit phase status', async () => {
-      const committee = await CommitteeUtils.createCommittee(
-        config.committee.consensus,
-        3,
-        web3.utils.sha3('dislocation'),
-        web3.utils.sha3('proposal'),
+    it('should fail if committee is in open phase status', async () => {
```
Comment (pgev) [inline test/committee/close_commit_phase.js:75]: As it has been decided to assert the condition in all other states, we are missing `CommitteeStatus::Closed` state here.

File: test/committee/cooldown_committee.js [hunk 8]
```diff
@@ -142,10 +127,10 @@ contract('Committee:cooldownCommittee', async (accounts) => {
       );
     });
 
-    it('should fail if committee is not in open state', async () => {
+    it('should fail if committee is in cooldown state', async () => {
```
Comment (pgev) [inline test/committee/cooldown_committee.js:129]: As it has been decided to assert the condition in all other states, we are missing `CommitteeStatus::Closed` state here.

**Review by pgev — CHANGES_REQUESTED** [review]
Thanks for changes.

Negative tests asserting correct states of functions became complicated with the current approach.
I would propose to introduce a MockCommittee contract that inherits from the actual one and adds a new function `setCommitteeState`. This will update the `committeeState` variable of the contract and tests would become much simpler and straightforward.

File: test/committee/cooldown_committee.js (hunk 8 above)
Comment (gulshanvasnani) [reply test/committee/cooldown_committee.js:129]: Done

File: test/committee/close_commit_phase.js (hunk 7 above)
Comment (gulshanvasnani) [reply test/committee/close_commit_phase.js:75]: Done

File: test/committee/challenge_committee.js (hunk 6 above)
Comment (gulshanvasnani) [reply test/committee/challenge_committee.js:80]: Done

File: test/committee/utils.js (hunk 1 above)
Comment (gulshanvasnani) [reply test/committee/utils.js:24]: It is used for *challengeCommittee* transaction.
The code was already there. I moved it to Utils
https://github.com/mosaicdao/mosaic-1/blob/develop/test/committee/challenge_committee.js#L54

File: test/committee/activate_committee.js (hunk 5 above)
Comment (gulshanvasnani) [reply test/committee/activate_committee.js:84]: Done

File: test/committee/utils.js (hunk 4 above)
Comment (gulshanvasnani) [reply test/committee/utils.js]: Done

File: test/committee/utils.js (hunk 3 above)
Comment (gulshanvasnani) [reply test/committee/utils.js]: Done

File: test/committee/utils.js (hunk 2 above)
Comment (gulshanvasnani) [reply test/committee/utils.js]: **this** is not needed. So removed it.

**Comment by gulshanvasnani** [comment]
As the implementation can change, so negative unit test cases is not required now.

**Pull request closed without merging**